File: sol/1718/1718.py
class Solution:
    def constructDistancedSequence(self, n: int) -> List[int]:
        m = 2 * n - 1
        res = [0] * m
        def fill(i, used):
            nonlocal res
            if i == m:
                return True
            if res[i] != 0:
                return fill(i + 1, used)
            for v in range(n, 1, -1):
                if used[v] or i + v >= m or res[i + v] != 0:
                    continue
                res[i] = v
                res[i + v] = v
                used[v] = True
                if fill(i + 1, used):
                    return True
                res[i] = 0
                res[i + v] = 0
                used[v] = False
            if not used[1]:
                res[i] = 1
                used[1] = True
                if fill(i + 1, used):
                    return True
                res[i] = 0
                used[1] = False
            return False
        fill(0, [False] * (n + 1))
        return res

File: sol/1718/1718.cpp
struct Solution {
    bool fill(int i, vector<int> &res, vector<bool> &used, int n, int m){
        if (i == m){
            return true;
        }
        if (res[i] != 0){
            return this->fill(i + 1, res, used, n, m);
        }
        for (int v = n; v > 1; --v){
            if (used[v] || (i + v >= m) || (res[i + v] != 0)){
                continue;
            }
            res[i] = v;
            res[i + v] = v;
            used[v] = true;
            if (this->fill(i + 1, res, used, n, m)){
                return true;
            }
            res[i] = 0;
            res[i + v] = 0;
            used[v] = false;
        }
        if (!used[1]){
            res[i] = 1;
            used[1] = true;
            if (this->fill(i + 1, res, used, n, m)){
                return true;
            }
            res[i] = 0;
            used[1] = false;
        }
        return false;
    }
    vector<int> constructDistancedSequence(int n) {
        vector<int> res(2 * n - 1, 0);
        vector<bool> used(n + 1, false);
        this->fill(0, res, used, n, 2 * n - 1);
        return res;
    }
};
